File: Chapter13/VB/Client Side/Order Entry/Router.vb
Imports System.Runtime.Remoting
Imports System.Runtime.Remoting.Services
Imports System.Runtime.Remoting.Channels
Imports System.Runtime.Remoting.Channels.Tcp

Imports Asap.Cars.CommonTypes
Imports Asap.Cars.OrderProcessor

Public Class Router
  Private _orderSystem As OrderProcessor.OrderSystem

  Public Sub New()
    EstablishConnectionToServer()
  End Sub

  Sub EstablishConnectionToServer()
    Try
      Dim configFile As String = System.Windows.Forms.Application.ExecutablePath + ".config"
      RemotingConfiguration.Configure(configFile)

      _orderSystem = DirectCast(Activator.GetObject(GetType(OrderProcessor.OrderSystem), "tcp://localhost:8011/AsapOrders"), OrderProcessor.OrderSystem)
    Catch ex As Exception
      Throw New Exception("Couldn't connect to Order Processing server. Details:\n\n" + ex.Message)
    End Try
  End Sub

  Public Sub SubmitOrder(ByVal theModel As String, ByVal theStyle As String, ByVal theColor As Color, ByVal theOptions As PricedItem())
    _orderSystem.SubmitOrder(theModel, theStyle, theColor, theOptions)
  End Sub

  Public Function GetModels() As String()
    Return _orderSystem.GetModels()
  End Function

  Public Function GetStyles(ByVal theModel As String) As String()
    Return _orderSystem.GetStyles(theModel)
  End Function

  Public Function GetColors(ByVal theModel As String, ByVal theStyle As String) As Color()
    Return _orderSystem.GetColors(theModel, theStyle)
  End Function

  ' each entry in the returned array is a PricedItem[]
  Public Function GetOptions(ByVal theModel As String, ByVal theStyle As String) As ArrayList
    Return _orderSystem.GetOptions(theModel, theStyle)
  End Function
End Class

File: Chapter13/C#/Client Side/Order Entry/Router.cs
using System;
using System.Drawing;
using System.ComponentModel;
using System.Collections;
using System.Runtime.Remoting;
using System.Runtime.Remoting.Services;
using System.Runtime.Remoting.Channels;
using System.Runtime.Remoting.Channels.Tcp;

using Asap.Cars.CommonTypes;
using Asap.Cars.OrderProcessor;

namespace Asap.Cars.OrderEntry
{
	/// <summary>
	/// Summary description for Router.
	/// </summary>
  public class Router
  {
    OrderProcessor.OrderSystem orderSystem;
    
	  public Router()
	  {
      EstablishConnectionToServer();
	  }

    void EstablishConnectionToServer()
    {
      try
      {
        string configFile = System.Windows.Forms.Application.ExecutablePath + ".config";
        RemotingConfiguration.Configure(configFile);

        orderSystem = (OrderProcessor.OrderSystem)Activator.GetObject(typeof(OrderProcessor.OrderSystem), "tcp://localhost:8011/AsapOrders");
      }
      catch (Exception ex)
      {
        throw new Exception("Couldn't connect to Order Processing server. Details:\n\n" + ex.Message);
      }
    }

    public void SubmitOrder(string theModel, string theStyle, Color theColor, PricedItem[] theOptions)
    {
      orderSystem.SubmitOrder(theModel, theStyle, theColor, theOptions);
    }

    public string[] GetModels()
    {
      return orderSystem.GetModels();
    }

    public string[] GetStyles(string theModel)
    {
      return orderSystem.GetStyles(theModel);
    }

    public Color[] GetColors(string theModel, string theStyle)
    {
      return orderSystem.GetColors(theModel, theStyle);
    }

    // each entry in the returned array is a PricedItem[]
    public ArrayList GetOptions(string theModel, string theStyle)
    {
      return orderSystem.GetOptions(theModel, theStyle);
    }
  }
}
